namespace SunamoCl.SunamoCmd.Essential;

public class CmdApp
{
    /// <summary>
    ///     Dont ask in console, load from Clipboard
    /// </summary>
    //public static bool loadFromClipboard = false;
    public static bool waitOnEnd = false;

    public static bool openAndWaitForChangeContentOfInputFile = true;


    /// <summary>
    ///     Create in class where are you calling method without A2 openVsCode
    /// </summary>
    /// <param name="myPositionsHtmlFile"></param>
    /// <param name="openVsCode"></param>
    public static
#if ASYNC
        async Task<string>
#else
    string
#endif
        WaitForSaving(string myPositionsHtmlFile, Func<string, Task> openVsCode)
    {
        Console.WriteLine($"Running WaitForSaving for {myPositionsHtmlFile}, openAndWaitForChangeContentOfInputFile: {openAndWaitForChangeContentOfInputFile}");

        if (openAndWaitForChangeContentOfInputFile)
        {
            await openVsCode(myPositionsHtmlFile);
            CL.WriteLine(
                $"Waiting for insert html to {Path.GetFileName(myPositionsHtmlFile)}, press enter to continue");
            CL.ReadLine();
        }

        Console.WriteLine($"Reading {myPositionsHtmlFile}");

        if (!File.Exists(myPositionsHtmlFile))
        {
            await File.WriteAllTextAsync(myPositionsHtmlFile, string.Empty);
            return string.Empty;
        }

        return
#if ASYNC
            await
#endif
                File.ReadAllTextAsync(myPositionsHtmlFile);
    }

    public static void WaitOnEnd()
    {
#if DEBUG
        if (waitOnEnd) CL.ReadLine();
#endif
    }

    public static void Init()
    {
        // Nevím zda je dobrý nápad. Když vznikne nechycená exception, dostane se do UnhandledExceptionTrapper() ale už to nemám v debuggeru VS. Možná to je jen špatným nastavením IDE.

        //AppDomain.CurrentDomain.UnhandledException += UnhandledExceptionTrapper;
    }

    /// <summary>
    /// Nevím zda je dobrý nápad. Když vznikne nechycená exception, dostane se do UnhandledExceptionTrapper() ale už to nemám v debuggeru VS. Možná to je jen špatným nastavením IDE.
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    internal static void UnhandledExceptionTrapper(object sender, UnhandledExceptionEventArgs e)
    {
        throw (Exception)e.ExceptionObject;

        //string dump = null;
        ////dump = YamlHelper.DumpAsYaml(e);
        ////dump = SunamoJsonHelper.SerializeObject(e, true);
        //dump = RH.DumpAsString(new DumpAsStringArgs { o = e, d = DumpProvider.Reflection });

        //ThisApp.Error(e.ExceptionObject.ToString());
        ////WriterEventLog.WriteToMainAppLog(dump, System.Diagnostics.EventLogEntryType.Error, Exc.CallingMethod());
    }

    public static void EnableConsoleLogging(bool v)
    {
        //if (v)
        //{
        //    // because method was called two times 
        //    ThisApp.StatusSetted -= ThisApp_StatusSetted;
        //    ThisApp.StatusSetted += ThisApp_StatusSetted;
        //}
        //else
        //{
        //    ThisApp.StatusSetted -= ThisApp_StatusSetted;
        //}
    }

    private static void ThisApp_StatusSetted(TypeOfMessageCl t, string message)
    {
        //TypedConsoleLogger.Instance.WriteLine(t, message);
    }

    /// <summary>
    ///     Alternatives are:
    ///     InitApp.SetDebugLogger
    ///     CmdApp.SetLogger
    ///     WpfApp.SetLogger
    /// </summary>
    public static void SetLogger()
    {
        //InitApp.Logger = ConsoleLoggerCmd.Instance;
        //InitApp.TemplateLogger = ConsoleTemplateLogger.Instance;
        //InitApp.TypedLogger = TypedConsoleLogger.Instance;
    }
}